internal/cnpi/plugin/client/cluster.go

```diff
@@ -92,6 +93,58 @@ func (data *data) MutateCluster(ctx context.Context, object client.Object, mutat
 	return nil
 }
 
+var (
+	errInvalidJSON      = errors.New("invalid json")
+	errSetClusterStatus = errors.New("set cluster status invokation failed")
+)
+
+func (data *data) SetClusterStatus(ctx context.Context, cluster client.Object) (map[string]string, error) {
+	contextLogger := log.FromContext(ctx)
+	serializedObject, err := json.Marshal(cluster)
+	if err != nil {
+		return nil, fmt.Errorf("while serializing %s %s/%s to JSON: %w",
+			cluster.GetObjectKind().GroupVersionKind().Kind,
+			cluster.GetNamespace(), cluster.GetName(),
+			err,
+		)
+	}
+
+	pluginStatuses := make(map[string]string)
+	for idx := range data.plugins {
+		plugin := data.plugins[idx]
+
+		if !slices.Contains(plugin.OperatorCapabilities(), operator.OperatorCapability_RPC_TYPE_SET_CLUSTER_STATUS) {
+			continue
+		}
+
+		pluginLogger := contextLogger.WithValues("pluginName", plugin.Name())
+		request := operator.SetClusterStatusRequest{
+			Cluster: serializedObject,
+		}
+
+		pluginLogger.Trace("Calling SetClusterStatus endpoint")
+		response, err := plugin.OperatorClient().SetClusterStatus(ctx, &request)
+		if err != nil {
+			pluginLogger.Error(err, "Error while calling SetClusterStatus")
+			return nil, fmt.Errorf("%w: %w", errSetClusterStatus, err)
+		}
+
+		if len(response.JsonStatus) == 0 {
+			contextLogger.Trace("json status is empty, skipping it", "pluginName", plugin.Name())
+			continue
+		}
```

Comment:
Does this prevent the plugin from clearing a previously set status?

Reply:
An empty string does not represent a valid JSON object.
The status can be cleared by passing `{}`.
Having merged that, we're going to validate it in `cnpg-i-machinery` and in the operator too.

Reply:
This PR contains the validation, line 136. In the original draft it was not very accurate but I improved it.